meta/rpc.go

```diff
@@ -0,0 +1,445 @@
+		var sz uint64
+		if err := binary.Read(conn, binary.BigEndian, &sz); err != nil {
+			return internal.RPCType_Error, nil, fmt.Errorf("read size: %s", err)
+		}
+
+		if sz == 0 {
+			return 0, nil, fmt.Errorf("invalid message size: %d", sz)
+		}
+
+		if sz >= MaxMessageSize {
+			return 0, nil, fmt.Errorf("max message size of %d exceeded: %d", MaxMessageSize, sz)
+		}
+
+		// Read request.
+		buf := make([]byte, sz)
+		if _, err := io.ReadFull(conn, buf); err != nil {
+			return internal.RPCType_Error, nil, fmt.Errorf("read request: %s", err)
+		}
+
+		// Determine the RPC type
+		rpcType := internal.RPCType(btou64(buf[0:8]))
+		buf = buf[8:]
+
+		r.traceCluster("recv %v request on: %v", rpcType, conn.RemoteAddr())
+		switch rpcType {
+		case internal.RPCType_FetchData:
+			var req internal.FetchDataRequest
+			if err := proto.Unmarshal(buf, &req); err != nil {
+				return internal.RPCType_Error, nil, fmt.Errorf("fetch request unmarshal: %v", err)
+			}
+			resp, err := r.handleFetchData(&req)
+			return rpcType, resp, err
+		case internal.RPCType_Join:
+			var req internal.JoinRequest
+			if err := proto.Unmarshal(buf, &req); err != nil {
+				return internal.RPCType_Error, nil, fmt.Errorf("join request unmarshal: %v", err)
+			}
+			resp, err := r.handleJoinRequest(&req)
+			return rpcType, resp, err
+		default:
+			return internal.RPCType_Error, nil, fmt.Errorf("unknown rpc type:%v", rpcType)
+		}
+	}()
+
+	// Handle unexpected RPC errors
+	if err != nil {
+		resp = &internal.ErrorResponse{
+			Header: &internal.ResponseHeader{
+				OK: proto.Bool(false),
+			},
+		}
+		typ = internal.RPCType_Error
+	}
+
+	// Set the status header and error message
+	if reply, ok := resp.(Reply); ok {
+		reply.GetHeader().OK = proto.Bool(err == nil)
+		if err != nil {
+			reply.GetHeader().Error = proto.String(err.Error())
+		}
+	}
+
+	r.sendResponse(conn, typ, resp)
+}
+
+func (r *RPC) sendResponse(conn net.Conn, typ internal.RPCType, resp proto.Message) {
+	// Marshal the response back to a protobuf
+	buf, err := proto.Marshal(resp)
+	if err != nil {
+		r.Logger.Printf("unable to marshal response: %v", err)
+		return
+	}
+
+	// Encode response back to connection.
+	if _, err := conn.Write(r.pack(typ, buf)); err != nil {
+		r.Logger.Printf("unable to write rpc response: %s", err)
+	}
+}
+
+func (r *RPC) sendError(conn net.Conn, msg string) {
+	r.traceCluster(msg)
+	resp := &internal.ErrorResponse{
+		Header: &internal.ResponseHeader{
+			OK:    proto.Bool(false),
+			Error: proto.String(msg),
+		},
+	}
+
+	r.sendResponse(conn, internal.RPCType_Error, resp)
+}
+
+// handleFetchData handles a request for the current nodes meta data
+func (r *RPC) handleFetchData(req *internal.FetchDataRequest) (*internal.FetchDataResponse, error) {
+	var (
+		b    []byte
+		data *Data
+		err  error
+	)
+
+	for {
+		data = r.store.cachedData()
+		if data.Index != req.GetIndex() {
+			b, err = data.MarshalBinary()
+			if err != nil {
+				return nil, err
+			}
+			break
+		}
```

Comment:
Not sure I follow, if the store index `data.Index` is higher than `req.GetIndex()`, then the L187 would be true and it would break out of the loop.